src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.SslProvider.cs

```diff
@@ -47,84 +47,83 @@ internal static void SetSslOptions(EasyRequest easy, ClientCertificateOption cli
-                                easy.SetCurlOption(Interop.Http.CURLoption.CURLOPT_SSL_VERIFYHOST, 0); // don't verify the peer cert's hostname
-                                // We don't change the SSL_VERIFYPEER setting, as setting it to 0 will cause
-                                // SSL and libcurl to ignore the result of the server callback.
-                            }
+                    // Register the callback with libcurl.  We need to register even if there's no user-provided
+                    // server callback and even if there are no client certificates, because we support verifying
+                    // server certificates against more than those known to OpenSSL.
+                    SetSslOptionsForSupportedBackend(easy, certProvider, userPointer);
+                }
+                else
+                {
+                    // Newer versions of OpenSSL, and other non-OpenSSL backends, do not currently support callbacks.
+                    // That means we'll throw a PNSE if a callback is required.
+                    SetSslOptionsForUnsupportedBackend(easy, certProvider);
+                }
+            }
 
-                            // The allowed SSL protocols will be set in the configuration callback.
-                            break;
-
-                        case CURLcode.CURLE_UNKNOWN_OPTION: // Curl 7.38 and prior
-                        case CURLcode.CURLE_NOT_BUILT_IN:   // Curl 7.39 and later
-                            // It's ok if we failed to register the callback if all of the defaults are in play
-                            // with relation to handling of certificates.  But if that's not the case, failing to 
-                            // register the callback will result in those options not being factored in, which is
-                            // a significant enough error that we need to fail.
-                            EventSourceTrace("CURLOPT_SSL_CTX_FUNCTION not supported: {0}", answer, easy: easy);
-                            if (certProvider != null ||
-                                easy._handler.ServerCertificateValidationCallback != null ||
-                                easy._handler.CheckCertificateRevocationList)
-                            {
-                                throw new PlatformNotSupportedException(
-                                    SR.Format(SR.net_http_unix_invalid_certcallback_option, CurlVersionDescription, CurlSslVersionDescription));
-                            }
+            private static void SetSslOptionsForSupportedBackend(EasyRequest easy, ClientCertificateProvider certProvider, IntPtr userPointer)
+            {
+                CURLcode answer = easy.SetSslCtxCallback(s_sslCtxCallback, userPointer);
+                switch (answer)
+                {
+                    case CURLcode.CURLE_OK:
+                        // We successfully registered.  If we'll be invoking a user-provided callback to verify the server
```

Comment:
I'm a bit confused by the comment: I thought the callback will be invoked internally regardless if the user provided a callback or not (if the public API is null) to perform the chain verification then the name verification.

Because if this isn't true: wouldn't we keep the host name validation enabled?